3.next - Fix Buffered query interactions with Collection

markstory opened the pull request:
`BufferedStatements` have a few problems with Collection as outlined in #12096. These changes solve that problem and allow BufferedStatements to be iterated multiple times. This required significant changes to `BufferedStatement` so I'm targeting 3.next for these changes.

markstory committed:
Fix some of the issues with buffered queries not buffering.

When a buffered query is added to a collection the IteratorIterator
constructor peels off the buffering decorator causing data to be lost
after the first collection operation.

Correctly buffering results requires another iterator that captures
results from the decorated statement.

markstory committed:
Fix implementation of BufferedStatement

Implement the Iterator interface directly. This allows the statement to
actually buffer results between iterations. It also solves the pierce
through problems that arise when statements are used with Collection.

markstory commented on tests/TestCase/Database/QueryTest.php:
```diff
@@ -2869,7 +2869,7 @@ public function testDecorateResults()
 
         $results = $query->decorateResults(null, true)->execute();
-        while ($row = $result->fetch('assoc')) {
+        while ($row = $results->fetch('assoc')) {
```
This was a mistake in the test code :blush:

lorenzo commented on src/Database/Statement/BufferedStatement.php:
```diff
@@ -14,54 +14,199 @@
+     */
+    public function closeCursor()
+    {
+        $this->statement->closeCursor();
```
Should we set the _allFetched variable here?

markstory replied:
I don't know.. I can add a test for iterating a closed statement.

lorenzo commented on src/Database/Statement/BufferedStatement.php:
```diff
@@ -73,25 +218,29 @@ public function execute($params = null)
+            $this->index += 1;
+
+            if ($row && $type === static::FETCH_TYPE_NUM) {
+                return array_values($row);
```
Clever!

markstory committed:
Make test conditional on sqlite.

Only sqlite requires userland buffering.

markstory committed:
Fix PHPStan warning.

markstory committed:
Rename property.

The TypeConverterTrait uses `_driver` and not `driver`.

lorenzo commented:
Are the failing tests for sql server relevant?

markstory commented:
@lorenzo I think they might be. I'm going to check tonight on my VM.

markstory committed:
Simplify BufferedStatement

Don't implement methods that can be simple proxies. Add a typehint to
the constructor to catch improper decorator usage earlier.

markstory committed:
Remove re-execution of buffered queries.

This solves incorrect results in SQLServer, as re-executing the query
would drop text columns for reasons I don't really understand. As
a side-effect loading associated data should be faster.

markstory commented:
The SQLServer errors were real failures, but I think I have them resolved now.

markstory commented on src/Database/Statement/BufferedStatement.php:
```diff
@@ -14,54 +14,176 @@
+     */
+    public function lastInsertId($table = null, $column = null)
+    {
+        return $this->_statement->lastInsertId($table, $column);
```
Yes. I'm a goof.

markstory committed:
Fix typo

markstory committed:
Attempt to fix constant namespacing issues.

Builds on travis are failing with undefined constants despite these
constants being part of PHP. Try to see if namespacing will help.

markstory committed:
Add magic property doc hint.
